### ISSUE TITLE
Signal to indicate preferred prioritization scheme

### ISSUE THREAD

**Issue by rmarx** [kazuho/draft-kazuho-httpbis-priority#50]
Since it looks like this draft will grow into the full new H3 proposal, I am posting this somewhat more tangential issue here as well. Let's dub it the "scheme selection signal". 

One of the things we keep seeing in our synthetic prioritization tests, is that the optimal scheme (scheme = round-robin, fifo, a specific combination of the two, see https://h3.edm.uhasselt.be/images/streams.png) can be **different per web page**. This is a problem, since for H2, each browser basically hard-codes a single scheme and servers are expected to follow it. This means some pages load much worse in some browser than others. Developers should thus ideally be able to tweak the browser (or server) behavior for certain use cases. 

The most practical solution we came up with in our [paper](https://h3.edm.uhasselt.be/) (since creating full prioritized lists of all resources, either manually or automatically, isn't very feasible), is to allow developers to select one of x different "schemes", indicating that scheme performs best for a particular page (e.g., for this page, you should use a weighted round-robin, for that one Patrick Meenan's structure). The optimal scheme can be determined practically by e.g., loading the pages offline using something like Webpagetest, and seeing if something sticks out positively (or negatively). 

As such, you would need a **second signal** (next to #49), indicating not how/which signals you wish to receive (e.g., dependency tree vs HTTP headers from this document), but rather **how to act on those signals.** For example, this document's approach allows servers much more leeway than H2's tree in how to actually interpret the prioritization signals. An urgency of 'prerequisite' could be interpreted as "send in full before the rest" but also as "add to Round-Robin with highest possible weight".  

To make this work, I envision that next to the "scheme selection signal", we would also have to define **a list of "default" schemes**, identifiable by a string name, and their globally expected behavior. I feel we might want such a list or at least some examples **either way** (even if we don't do this second signal), to guide implementers on how to actually implement the new approach (how can/should a server translate 'urgency' and 'progressive' into bandwidth distribution). I fear that otherwise, in combination with the very unspecified quic-layer prioritization, this work will have sub-optimal impact, as implementers will choose the most direct/naive mapping of the signals to bandwidth-level behavior. 

What such a "scheme selection signal" and list of schemes would look like is a bit unclear to myself, as there are several caveats, e.g., what when coalescing, loading multiple pages back-to-back over a connection (makes using SETTING difficult), how many schemes would you describe, etc. I do think we need a signal (i.e., we cannot just make this a server-side config), since an origin server might want to influence a CDN (see what Cloudflare is doing with H2 priorities atm) and on H2, the browser might need to be the one tweaking their dependency tree to achieve the preferred behavior. 

---

TL;DR

- I feel we need a list of example "prioritization schemes" that show developers how to best interpret the signals defined in this draft
- If we have this list, we can allow developers to indicate their "preferred scheme" for a specific page / site, if they know which one performs best. For this, we would need a secondary signal. 
- Browsers and servers can then choose to implement multiple different schemes and switch between them as signaled. 

Is there any interest in this concept from the team? Is this something we might wish to tackle now or for "V2"? 

If the signal itself is too complex / too early, do we indeed want a "list of example schemes" to guide implementers? If yes, should that be in this document?

**Comment by LPardue** [comment]
One challenge is that it seems like you want to the ability for a document to define properties related to the underlying connection(s) that will be used to fetch resources. I can't think of any examples where this is allowed today. I fear this breaks abstractions and the models built around them. It's not a perfect examplue but even the Fetch API has indirect access to many Connection-level properties.

**Comment by LPardue** [comment]
@rmarx I think some of the concepts in this issue relate to other open issues (e.g. #95, #86). Is it possible to extract some simpler issue that we can consider or fix indenpendently, or is this subsumed by other tickets. If its the latter we can keep this ticket open for tracking purposes. If its neither then maybe we should consider closing this one.

**Comment by LPardue** [comment]
@rmarx please open a new issue at https://github.com/httpwg/http-extensions if you feel compelled.
